### PR TITLE
Wrap up performance evaluation

### PR DESCRIPTION
**Pull request by nkarag**
Added elapsed time and memory consumed measurements.

### CODE REVIEW

File: executable-spec/bench/micro-benchmarking/Main.hs [hunk 1]
```diff
@@ -13,18 +13,18 @@ main :: IO ()
 main = do
   let
       !tallyData2 =
-        createTallyData constants (NumberOfParticipants 10)      (NumberOfConcurrentUPs 10)
+        createTallyData constants (NumberOfParticipants 100)      (NumberOfConcurrentUPs 1)
       !tallyData3 =
-        createTallyData constants (NumberOfParticipants 100)     (NumberOfConcurrentUPs 10)
+         createTallyData constants (NumberOfParticipants 1000)     (NumberOfConcurrentUPs 1)
```
Comment (dnadales) [inline executable-spec/bench/micro-benchmarking/Main.hs]: Check the alignment :)

File: formal-spec/decentralized-updates.tex [hunk 2]
```diff
@@ -50,6 +51,28 @@
 
 \newcommand{\nnote}[1]{{\color{blue}\small Nikos: #1}}
 
+
+\lstdefinestyle{mystyle}{
+%    backgroundcolor=\color{backcolour},
```
Comment (dnadales) [inline formal-spec/decentralized-updates.tex]: Maybe remove things that are commented out?

File: formal-spec/measurements.tex [hunk 3]
```diff
@@ -141,8 +143,9 @@ \section{Impact on performance}
       ylabel={Usage percentage},
       ymin=0.0,
       ymax=250.0,
-      ymode=linear,
-      ytick={5.0, 20.0, 50.0, 100.0, 150.0, 200.0, 250.0},
+      ymode=log, %linear,
```
Comment (dnadales) [inline formal-spec/measurements.tex]: Why the change in the scale?

Also remove the code that isn't needed.

File: formal-spec/measurements.tex [hunk 4]
```diff
@@ -156,7 +159,7 @@ \section{Impact on performance}
 \end{figure}
 
 We can see that the impact on the system's performance is negligible even when
-we consider $100,000$ participants.
+we consider $100,000$ participants. Moreover, we see that the usage consumption percentage scales linearly in the number of participants, i.e., a 10 times increase in the number of participants will only increase 10 times the required usage percent. Also, if we double the throughput, then we can process a double amount of participants workload at the same time.
```
Comment (dnadales) [inline formal-spec/measurements.tex]: Limit the line length to 80 characters. It's easier to review :)

File: formal-spec/measurements.tex [hunk 5]
```diff
@@ -174,6 +177,102 @@ \section{Impact on performance}
 %% the section above.
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
+\subsection{Processing Time and Memory Consumption}
```
Comment (dnadales) [inline formal-spec/measurements.tex]: Please wrap everything to 80 chars :pray:

File: formal-spec/measurements.tex [hunk 6]
```diff
@@ -174,6 +177,102 @@ \section{Impact on performance}
...
+  if stakeThreshold adversarialStakeRatio (totalStake stakeDistribution)
+     <
+     stakeOfKeys votingKeys stakeDistribution
+  then Just result
+  else Nothing
+  where
+    votingKeys = Map.filter (== confidence) ballot
+\end{lstlisting}
+
+\lstinline{Map.filter} is $O(n)$ \footnote{\href{url}{http://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Map-Strict.html\#g:25}}, so \lstinline{votingKeys} is $O(n)$, where $n$ is the number of ballots, which as we have said coincides to the number of participants. At this point, we have a single pass (loop) over $n$ ballots.
```
Comment (dnadales) [inline formal-spec/measurements.tex]: The foot note makes it appear similar to `O(n^2)`. I'd place it after `Map.filter` to avoid confusion.

File: formal-spec/measurements.tex [hunk 7]
```diff
@@ -174,6 +177,102 @@ \section{Impact on performance}
 %% the section above.
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
+\subsection{Processing Time and Memory Consumption}
+Clearly the most processing intesive task of the update mechanism is the tally phase. It is the phase where all the collected votes are counted in order to reach at a decision for a specific proposal. 
+
+We start with a theoretical time complexity analysis where we assume a worst-case scenario, where we have $n$ participants that all of them vote by submiting a single vote. Also we assume that we have a single proposal, so that within a voting period, the number $n$ of participants coincides to the number of submitted ballots.
+
+In the following we try to break up the operations during the tally phase. In the heart of the tally phase lies the following function call, which is called three times sequencially for each proposal.
```
Comment (dnadales) [inline formal-spec/measurements.tex]: I wouldn't mention that it is called three times. We implemented this in this way, but we will likely change this to make a single pass.

File: formal-spec/measurements.tex [hunk 8]
```diff
@@ -174,6 +177,102 @@ \section{Impact on performance}
...
+      ]
+      \addplot[color=black] table {participants-vs-elapsed_time.dat};
+    \end{axis}
+  \end{tikzpicture}
+
+  \caption{Worst case scenario analysis for tally phase processing time}
+  \label{fig:eltime-vs-participants}
+\end{figure}
+
+In Figure \ref{fig:eltime-vs-participants}, we see the elapsed time of the tally phase to scale linearly in the number of participants. In addition, we see that it takes almost one tenth of a second to process the votes of $1$ million participants. These results correspond to a no-parallel execution of the tally algorithm on a i7 CPU laptop with 32GB of RAM.
```
Comment (dnadales) [inline formal-spec/measurements.tex]: I'd say "the elapsed time **increases** linearly"

File: executable-spec/bench/micro-benchmarking/Main.hs (hunk 1 above)
Comment (nkarag) [reply executable-spec/bench/micro-benchmarking/Main.hs]: OK

File: formal-spec/decentralized-updates.tex (hunk 2 above)
Comment (nkarag) [reply formal-spec/decentralized-updates.tex]: I thought that maybe we can leave the comments that correspond to a few specific key-words/attributes of the coding-list style definition, so as not to have to look for them, in the case we need them.

File: formal-spec/measurements.tex (hunk 3 above)
Comment (nkarag) [reply formal-spec/measurements.tex]: I think that the log scale depicts better that an increase by 10 times in the number of participants , increases by 10x the usage percentage. Log scales show relative values (increase by 10x) instead of absolute ones and (at least to me) it is more intuitive.

File: formal-spec/measurements.tex (hunk 4 above)
Comment (nkarag) [reply formal-spec/measurements.tex]: OK

File: formal-spec/measurements.tex (hunk 5 above)
Comment (nkarag) [reply formal-spec/measurements.tex]: OK

File: formal-spec/measurements.tex (hunk 6 above)
Comment (nkarag) [reply formal-spec/measurements.tex]: OK

File: formal-spec/measurements.tex (hunk 8 above)
Comment (nkarag) [reply formal-spec/measurements.tex]: OK

File: formal-spec/measurements.tex (hunk 7 above)
Comment (nkarag) [reply formal-spec/measurements.tex]: OK
